lib/fog/compute/kubevirt.rb

```diff
@@ -84,26 +137,162 @@ def initialize(options={})
           # Kubeclient needs different client objects for different API groups. We will keep in this hash the
           # client objects, indexed by API path/version.
           @clients = {}
+        end
+
+        def virt_supported?
+          virt_enabled = false
 
-          @client = kubevirt_client()
+          begin
+            api_versions = kubevirt_client.api["versions"]
+            api_versions.each do |ver|
+              if ver["groupVersion"]&.start_with?(KUBEVIRT_GROUP)
+                virt_enabled = true
+              end
+            end
+          rescue => err
+            # we failed to communicate or to evaluate the version format
+            @log.warn("Failed to detect kubevirt on provider with error: #{err.message}")
+          end
 
-          # TODO expect a specific token
-          @oc_client = openshift_client()
-          @kube_client = kube_client()
+          virt_enabled
         end
 
-        private
+        def valid?
+          kube_client.api_valid?
+        end
 
-        def client
-          @client
+        class WatchWrapper
+
+          def initialize(watch, mapper)
+            @watch = watch
+            @mapper = mapper
+          end
+
+          def each
+            @watch.each do |notice|
+              yield @mapper.call(notice)
+            end
+          end
+
+          def finish
+            @watch.finish
+          end
         end
 
-        def oc_client
-          @oc_client
+        #
+        # Returns a watcher for nodes.
+        #
+        # @param opts [Hash] A hash with options for the watcher.
+        # @return [WatchWrapper] The watcher.
+        #
+        def watch_nodes(opts = {})
+          mapper = Proc.new do |notice|
+            if notice.object.kind == 'Node'
```

Comment:
will this work?
node = OpenStruct.new(Node.parse(notice.object)) if notice.object.kind == 'Node'
node ||= OpenStruct.new

I have to say I am not sure what this does. The proc does not seem to depend on opts, so why do we need to create a proc dynamically, why cant we just have a method that does this?

Reply:
We need to run this proc per notification we receive in the stream. If the notification kind is different than it means we need to restart watches (kubernetes mechanism). I will check whether your suggestion would work.